lib/bolt_ext/puppetdb_inventory.rb

```diff
@@ -0,0 +1,191 @@
+        if File.exist?(filename)
+          JSON.parse(File.read(filename)).fetch('puppetdb', {})
+        # Only try to use the default token if it exists, so we don't override
+        # cert/key with a non-existent token
+        elsif File.exist?(DEFAULT_TOKEN)
+          { 'token' => DEFAULT_TOKEN }
+        else
+          {}
+        end
+      end
+
+      def [](key)
+        @settings[key]
+      end
+
+      def validate
+        unless @settings['server_urls']
+          raise "server_urls must be specified in the config file or with --url"
+        end
+        unless @settings['cacert']
+          raise "cacert must be specified in the config file or with --cacert"
+        end
+        if !@settings['token'] && !@settings['cert'] && !@settings['key']
+          raise "token or cert/key must be provided"
+        end
+        if (@settings['cert'] && !@settings['key']) ||
+           (!@settings['cert'] && @settings['key'])
+          raise "cert and key must be specified together"
+        end
+      end
+    end
+
+    class CLI
+      def initialize(args)
+        @args = args
+        @cli_opts = {}
+        @parser = build_parser
+      end
+
+      def build_parser
+        parser = OptionParser.new('') do |opts|
+          opts.on('--cacert CACERT', "Path to the CA certificate") do |cacert|
+            @cli_opts['cacert'] = cacert
+          end
+          opts.on('--cert CERT', "Path to the certificate") do |cert|
+            @cli_opts['cert'] = cert
+          end
+          opts.on('--key KEY', "Path to the private key") do |key|
+            @cli_opts['key'] = key
+          end
+          opts.on('--token-file TOKEN', "Path to the token file") do |token|
+            @cli_opts['token'] = token
+          end
+          opts.on('--url URL', "The URL of the PuppetDB server to connect to") do |url|
+            @cli_opts['server_urls'] = [url]
+          end
+          opts.on('--file FILE', "Path to the inventory file to use") do |file|
+            @inventory_file = File.expand_path(file)
+          end
+          opts.on('-h', '--help', "Display help") do |_|
+            @show_help = true
+          end
+        end
+        parser.banner = <<-BANNER
+  Usage: bolt-inventory-pdb --file <inventory-file> [--url <url>] [auth-options]
+
+  Populate the nodes in an inventory file based on PuppetDB queries.
+
+        BANNER
+        parser
+      end
+
+      def run
+        @parser.permute(@args)
+
+        if @show_help
+          puts @parser.help
+          return 0
+        end
+
+        unless @inventory_file
+          raise "--file is a required option"
+        end
+
+        config = Config.new(@cli_opts)
+        @puppetdb_client = Client.new(config)
+
+        unless File.readable?(@inventory_file)
+          raise "Can't read the inventory file #{@inventory_file}"
+        end
+
+        inventory = YAML.load_file(@inventory_file)
+        resolve_group(inventory)
+
+        puts inventory.to_yaml
```

Comment:
spare puts (not sure if this is final or not)

Reply:
Does the script write the result to stdout? It'd be nice if that were clear in the help output.

Reply:
With stdout, I can't do an in-place update of my inventory file. Trying to do so ends up overwriting the file before it's read.

Reply:
This `puts` was intended, to write to stdout. But I didn't realize that still wouldn't let you update it in place. I can add an option for output file.